src/modules/SX126x/SX126x.cpp

```diff
@@ -443,7 +443,91 @@ int16_t SX126x::startTransmit(uint8_t* data, size_t len, uint8_t addr) {
 }
 
 int16_t SX126x::startReceive(uint32_t timeout) {
-  // set DIO mapping
+  int16_t state = startReceiveCommon();
+  if(state != ERR_NONE) {
+    return(state);
+  }
+
+  // set mode to receive
+  state = setRx(timeout);
+
+  return(state);
+}
+
+int16_t SX126x::startReceiveDutyCycle(uint32_t rxPeriod_us, uint32_t sleepPeriod_us)
+{
+  // datasheet claims time to go to sleep is ~500us, same to wake up.
+  // compensate for that 1ms + tcxo delay:
+  uint32_t transitionTime = _tcxoDelay_us + 1000;
+  if (sleepPeriod_us < transitionTime + 16) {
+    return(ERR_INVALID_SLEEP_PERIOD);
+  }
+  sleepPeriod_us -= transitionTime;
+
+  // divide by 15.625
+  uint32_t rxPeriodRaw = (rxPeriod_us * 8) / 125;
+  uint32_t sleepPeriodRaw = (sleepPeriod_us * 8) / 125;
+
+  // 24 bit limit:
+  if ((rxPeriodRaw & 0xFF000000) || (sleepPeriodRaw & 0xFF000000)) {
+    return(ERR_OVERFLOW);
+  }
+
+  int16_t state = startReceiveCommon();
+  if (state != ERR_NONE) {
+    return(state);
+  }
+
+  byte data[6] = {
+    (rxPeriodRaw >> 16) & 0xFF,    (rxPeriodRaw >> 8) & 0xFF,    rxPeriodRaw & 0xFF,
+    (sleepPeriodRaw >> 16) & 0xFF, (sleepPeriodRaw >> 8) & 0xFF, sleepPeriodRaw & 0xFF
+  };
+  return(SPIwriteCommand(SX126X_CMD_SET_RX_DUTY_CYCLE, data, 6));
+}
+
+int16_t SX126x::startReceiveDutyCycleAuto(uint16_t senderPreambleLength, uint16_t minSymbols)
+{
+  if (senderPreambleLength == 0) {
+    senderPreambleLength = _preambleLength;
+  }
+
+  // worst case is that the sender starts transmiting when we're just less than minSymbols from going back to sleep.
+  // in this case, we don't catch minSymbols before going to sleep,
+  // so we must be awake for at least that long before the sender stops transmitting.
+  uint16_t sleepSymbols = senderPreambleLength - 2 * minSymbols;
+
+  // if we're not to sleep at all, just use the standard startReceive.
+  if (2 * minSymbols > senderPreambleLength) {
+    return(startReceive());
+  }
+
+  uint32_t symbolLength_us = ((uint32_t)(10 * 1000) << _sf) / (10 * _bwKhz);
```

Comment:
I'm not entirely sure this is correct, since the symbol duration equation should be: Ts = (2^SF)/BW

Reply:
The 1000 is because symbolLength is in us and bw is in kHz:
Ts = (2^SF)/Bw.
Ts * 1E6 = (2^SF)/Bw * 1E6
Ts(us) = (2^SF)/(Bw(kHz) * 1000) * 1E6
Ts(us) = (2^SF)/(Bw(kHz)) * 1E3
Ts(us) = (1E3 * 2^SF) / BwKHz

Then it multiplies _bwKhz by 10 because it might be as low as 7.8. It was supposed to be cast to integer before the division, looks like I missed that. Anyway, multiply the numerator and denominator by the same number and the equation is unchanged, so that's why there's the 10 in the top too. I guess that could be left out and it can just work in FP.

Reply:
Checks out, thanks. I was confused as to how did the 10*1000 got there.